pkg/server/server.go

```diff
@@ -0,0 +1,159 @@
+	"istio.io/istio/security/pkg/server/ca/authenticate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cert-manager/istio-csr/cmd/app/options"
+	"github.com/cert-manager/istio-csr/pkg/util"
+)
+
+const (
+	IdentitiesAnnotationKey = "istio.cert-manager.io/identities"
+)
+
+// Server is the implementation of the istio CreateCertificate service
+type Server struct {
+	log *logrus.Entry
+
+	client cmclient.CertificateRequestInterface
+	auther authenticate.Authenticator
+
+	issuerRef   cmmeta.ObjectReference
+	preserveCRs bool
+}
+
+func New(log *logrus.Entry, cmOptions *options.CertManagerOptions, kubeOptions *options.KubeOptions) *Server {
+	return &Server{
+		log:         log.WithField("module", "certificate_provider"),
+		client:      kubeOptions.CMClient,
+		auther:      kubeOptions.Auther,
+		issuerRef:   cmOptions.IssuerRef,
+		preserveCRs: cmOptions.PreserveCRs,
+	}
+}
+
+// Run is a blocking func that will run the client facing certificate service
+func (s *Server) Run(ctx context.Context, tlsConfig *tls.Config, listenAddress string) error {
+	// Setup the grpc server using the passed TLS config
+	creds := credentials.NewTLS(tlsConfig)
+	grpcServer := grpc.NewServer(grpc.Creds(creds))
+
+	// listen on the configured address
+	listener, err := net.Listen("tcp", listenAddress)
+	if err != nil {
+		return fmt.Errorf("failed to listen %s: %v", listenAddress, err)
+	}
+
+	// register certificate service grpc API
+	securityapi.RegisterIstioCertificateServiceServer(grpcServer, s)
+
+	// handle termination gracefully
+	go func() {
+		<-ctx.Done()
+		s.log.Info("shutting down grpc server")
+		grpcServer.GracefulStop()
+		s.log.Info("grpc server stopped")
+	}()
+
+	s.log.Infof("grpc serving on %s", listener.Addr())
+
+	return grpcServer.Serve(listener)
+}
+
+// CreateCertificate is the istio grpc API func, to authenticate, authorize,
+// and sign CSRs requests from istio clients.
+func (s *Server) CreateCertificate(ctx context.Context, icr *securityapi.IstioCertificateRequest) (*securityapi.IstioCertificateResponse, error) {
+	// authn incoming requests, and build concatenated identities for labelling
+	identities, ok := s.authRequest(ctx, []byte(icr.Csr))
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "request authenticate failure")
+	}
+
+	// Build cert-manager CertificateRequest based on the configured issuer
+	cr := &cmapi.CertificateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			// Random non-conflicted name
+			GenerateName: "istio-",
+			Annotations: map[string]string{
+				// Label identities to resource for auditing
+				IdentitiesAnnotationKey: identities,
+			},
+		},
+		Spec: cmapi.CertificateRequestSpec{
+			Duration: &metav1.Duration{
+				// Add during which was requested from the client.
+				// TODO: We should have a configurable maximum that the duration can
+				// be. Take smaller of the two.
+				Duration: time.Duration(icr.ValidityDuration) * time.Second,
```

Comment:
Is this appropriate?

Reply:
Given that the agent could function as a local CA, it makes sense to have a flag that configures the maximum possible certificate duration